components/properties/src/props.rs

```diff
@@ -41,6 +49,147 @@ enum EnumeratedProperty {
+///
+/// # Example
+///
+/// ```
+/// use icu::properties::GeneralCategory;
+///
+/// let lookup = GeneralCategory::lookup_values(&icu_testdata::unstable())
+///                  .expect("The data should be valid");
+/// // short name for value
+/// assert_eq!(lookup.get_strict("Lu"), Some(GeneralCategory::UppercaseLetter));
+/// assert_eq!(lookup.get_strict("Pd"), Some(GeneralCategory::DashPunctuation));
+/// // long name for value
+/// assert_eq!(lookup.get_strict("Uppercase_Letter"), Some(GeneralCategory::UppercaseLetter));
+/// assert_eq!(lookup.get_strict("Dash_Punctuation"), Some(GeneralCategory::DashPunctuation));
+/// // name has incorrect casing
+/// assert_eq!(lookup.get_strict("dashpunctuation"), None);
+/// // loose matching of name
+/// assert_eq!(lookup.get_loose("dash-punctuation"), Some(GeneralCategory::DashPunctuation));
+/// // fake property
+/// assert_eq!(lookup.get_strict("Animated_Gif"), None);
+/// ```
+pub struct PropertyValueNameToEnumMapper<T> {
+    map: DataPayload<ErasedNameToEnumMapV1Marker>,
+    marker: PhantomData<fn() -> T>,
+}
+
+impl<T: TrieValue> PropertyValueNameToEnumMapper<T> {
+    /// Get the property value as a u16, doing a strict search looking for
+    /// names that match exactly
+    #[inline]
+    pub fn get_strict_u16(&self, name: &str) -> Option<u16> {
+        get_strict_u16(&self.map, name)
+    }
+
+    /// Get the property value as a `T`, doing a strict search looking for
+    /// names that match exactly
+    #[inline]
+    pub fn get_strict(&self, name: &str) -> Option<T> {
+        self.get_strict_u16(name)
+            .and_then(|u| T::try_from_u32(u as u32).ok())
+    }
+
+    /// Get the property value as a u16, doing a loose search looking for
+    /// names that match case-insensitively, ignoring ASCII hyphens, underscores, and
+    /// whitespaces.
+    #[inline]
+    pub fn get_loose_u16(&self, name: &str) -> Option<u16> {
+        get_loose_u16(&self.map, name)
+    }
+
+    /// Get the property value as a `T`, doing a loose search looking for
+    /// names that match case-insensitively, ignoring ASCII hyphens, underscores, and
+    /// whitespaces.
+    #[inline]
+    pub fn get_loose(&self, name: &str) -> Option<T> {
+        get_loose_u16(&self.map, name).and_then(|u| T::try_from_u32(u as u32).ok())
+    }
+
+    /// Construct a new one from loaded data
+    ///
+    /// Typically it is preferable to use methods on individual property value types
+    /// (like [`Script::lookup_values()`]) instead.
+    pub fn from_data<M>(data: DataPayload<M>) -> Self
+    where
+        M: DataMarker<Yokeable = PropertyValueNameToEnumMapV1<'static>>,
+    {
+        Self {
+            map: data.map_project(|m, _| m),
+            marker: PhantomData,
+        }
+    }
+}
+
+/// Avoid monomorphizing multiple copies of this function
+fn get_strict_u16(map: &DataPayload<ErasedNameToEnumMapV1Marker>, name: &str) -> Option<u16> {
```

Comment:
nit: call the parameter `payload` to not end up with `map.get().map`. Also on `loose`